overlord/snapshotstate/snapshotmgr.go

```diff
@@ -46,27 +48,90 @@ var (
 	backendCheck         = (*backend.Reader).Check
 	backendRevert        = (*backend.RestoreState).Revert // ditto
 	backendCleanup       = (*backend.RestoreState).Cleanup
+
+	snapshotExpirationLoopInterval = time.Hour * 24 // interval between forgetExpiredSnapshots runs as part of Ensure()
 )
 
 // SnapshotManager takes snapshots of active snaps
-type SnapshotManager struct{}
+type SnapshotManager struct {
+	state *state.State
+
+	lastForgetExpiredSnapshotTime time.Time
+}
 
 // Manager returns a new SnapshotManager
 func Manager(st *state.State, runner *state.TaskRunner) *SnapshotManager {
+	delayedCrossMgrInit()
+
 	runner.AddHandler("save-snapshot", doSave, doForget)
 	runner.AddHandler("forget-snapshot", doForget, nil)
 	runner.AddHandler("check-snapshot", doCheck, nil)
 	runner.AddHandler("restore-snapshot", doRestore, undoRestore)
 	runner.AddCleanup("restore-snapshot", cleanupRestore)
 
-	manager := &SnapshotManager{}
+	manager := &SnapshotManager{
+		state: st,
+	}
 	snapstate.AddAffectedSnapsByAttr("snapshot-setup", manager.affectedSnaps)
 
 	return manager
 }
 
 // Ensure is part of the overlord.StateManager interface.
-func (SnapshotManager) Ensure() error { return nil }
+func (mgr *SnapshotManager) Ensure() error {
+	// process expired snapshots once a day.
+	if time.Now().After(mgr.lastForgetExpiredSnapshotTime.Add(snapshotExpirationLoopInterval)) {
+		return mgr.forgetExpiredSnapshots()
+	}
+	return nil
+}
+
+func (mgr *SnapshotManager) forgetExpiredSnapshots() error {
+	mgr.state.Lock()
+	defer mgr.state.Unlock()
+
+	sets, err := expiredSnapshotSets(mgr.state, time.Now())
+	if err != nil {
+		return fmt.Errorf("internal error: cannot determine expired snapshots: %v", err)
+	}
+
+	if len(sets) == 0 {
+		return nil
+	}
+
+	err = backendIter(context.TODO(), func(r *backend.Reader) error {
+		// forget needs to conflict with check and restore
+		if err := checkSnapshotTaskConflict(mgr.state, r.SetID, "check-snapshot", "restore-snapshot"); err != nil {
+			// there is a conflict, do nothing and we will retry this set on next Ensure().
+			return nil
+		}
+		if sets[r.SetID] {
+			delete(sets, r.SetID)
+			// remove from state first: in case removeSnapshotState succeeds but osRemove fails we will never attempt
+			// to automatically remove this snapshot again and will leave it on the disk (so the user can still try to remove it manually);
+			// this is better than the other way around where a failing osRemove would be retried forever because snapshot would never
+			// leave the state.
+			if rmerr := removeSnapshotState(mgr.state, r.SetID); rmerr != nil {
+				return fmt.Errorf("internal error: cannot remove state of snapshot set %d: %v", r.SetID, rmerr)
+			}
+			if rmerr := osRemove(r.Name()); rmerr != nil {
+				return fmt.Errorf("cannot remove snapshot file %q: %v", r.Name(), rmerr)
```

Comment:
@chipaca @stolowski  does this merit a warning now that we have them?

Reply:
I had a short chat about that with @chipaca and yes, but it's probably a material for a followup PR because that's not the only instance where warnings could be used in snapshotstate.